backend/drm/renderer.c

```diff
@@ -266,46 +181,242 @@ static struct wlr_texture *get_tex_for_bo(struct wlr_drm_renderer *renderer,
 	struct wlr_drm_format_set *format_set =
 		with_modifiers ? &plane->formats : NULL;
 
+	drm_plane_fini_surface(plane);
+
 	if (!drm->parent) {
 		return init_drm_surface(&plane->surf, &drm->renderer, width, height,
-			format, format_set, GBM_BO_USE_SCANOUT);
+			format, format_set, flags | GBM_BO_USE_SCANOUT);
 	}
 
 	if (!init_drm_surface(&plane->surf, &drm->parent->renderer,
-			width, height, format, NULL, GBM_BO_USE_LINEAR)) {
+			width, height, format, NULL,
+			flags | GBM_BO_USE_LINEAR)) {
 		return false;
 	}
 
 	if (!init_drm_surface(&plane->mgpu_surf, &drm->renderer,
-			width, height, format, format_set, GBM_BO_USE_SCANOUT)) {
+			width, height, format, format_set,
+			flags | GBM_BO_USE_SCANOUT)) {
 		finish_drm_surface(&plane->surf);
 		return false;
 	}
 
 	return true;
 }
+
+void drm_fb_clear(struct wlr_drm_fb *fb) {
+	switch (fb->type) {
+	case WLR_DRM_FB_TYPE_NONE:
+		assert(!fb->bo);
+		break;
+	case WLR_DRM_FB_TYPE_SURFACE:
+		gbm_surface_release_buffer(fb->surf->gbm, fb->bo);
+		break;
+	case WLR_DRM_FB_TYPE_WLR_BUFFER:
+		gbm_bo_destroy(fb->bo);
+		wlr_buffer_unref(fb->wlr_buf);
+		fb->wlr_buf = NULL;
+		break;
+	}
+
+	fb->type = WLR_DRM_FB_TYPE_NONE;
+	fb->bo = NULL;
+
+	if (fb->mgpu_bo) {
+		assert(fb->mgpu_surf);
+		gbm_surface_release_buffer(fb->mgpu_surf->gbm, fb->mgpu_bo);
+		fb->mgpu_bo = NULL;
+		fb->mgpu_surf = NULL;
+	}
+}
+
+bool drm_fb_lock_surface(struct wlr_drm_fb *fb, struct wlr_drm_surface *surf) {
+	drm_fb_clear(fb);
+
+	if (!wlr_egl_swap_buffers(&surf->renderer->egl, surf->egl, NULL)) {
+		wlr_log(WLR_ERROR, "Failed to swap buffers");
+		return false;
+	}
+
+	fb->bo = gbm_surface_lock_front_buffer(surf->gbm);
+	if (!fb->bo) {
+		wlr_log(WLR_ERROR, "Failed to lock front buffer");
+		return false;
+	}
+
+	fb->type = WLR_DRM_FB_TYPE_SURFACE;
+	fb->surf = surf;
+	return true;
+}
+
+static uint32_t strip_alpha_channel(uint32_t format) {
+	switch (format) {
+	case DRM_FORMAT_ARGB8888:
+		return DRM_FORMAT_XRGB8888;
+	default:
+		return DRM_FORMAT_INVALID;
+	}
+}
+
+bool drm_fb_import_wlr(struct wlr_drm_fb *fb, struct wlr_drm_renderer *renderer,
+		struct wlr_buffer *buf, int req_width, int req_height,
+		struct wlr_drm_format_set *set) {
+	struct wlr_dmabuf_attributes attribs;
+
+	drm_fb_clear(fb);
+
+	if (!wlr_buffer_get_dmabuf(buf, &attribs)) {
+		return false;
+	}
+
+	/* TODO: Y-flipped is possible if adverstised by DRM properties */
+	if (attribs.flags != 0) {
+		return false;
+	}
+
+	if (attribs.width != req_width || attribs.height != req_height) {
+		return false;
+	}
+
+	if (!wlr_drm_format_set_has(set, attribs.format, attribs.modifier)) {
+		// The format isn't supported by the plane. Try stripping the alpha
+		// channel, if any.
+		uint32_t format = strip_alpha_channel(attribs.format);
+		if (wlr_drm_format_set_has(set, format, attribs.modifier)) {
+			attribs.format = format;
+		} else {
+			return false;
+		}
+	}
+
+	if (attribs.modifier != DRM_FORMAT_MOD_INVALID) {
```

Comment:
Can we only also use this code-path if `attribs.n_planes > 1 || attribs.offset[0] != 0`, like Weston?